### PR TITLE
[ml] Fix un-initialized centroids bug (k-means)

### PR DESCRIPTION
**Pull request by Yuki992411**
I added a function that initializes centroids in kmeans.

### CODE REVIEW

**Comment by mvieth** [comment]
How about putting the for-loop inside `computeCentroids`, at line 98? That would in my opinion make logically the most sense: set centroids to zero - sum up points - divide by number of points. Otherwise the function `computeCentroids` would still have the requirement that the centroids are initialized to zero before calling the function.

**Comment by Yuki992411** [comment]
Thank you for your feedback.
Oh, I missed it. It's more natural to add initialization at line 98.
I made a code with low maintainability because I tried to reduce the number of initializations. 
As you pointed out, I will correct it.

**Review by mvieth — APPROVED** [review]
Thanks!

**Review by larshg — APPROVED** [review]

**Comment by larshg** [comment]
Should we make a issue about creating a unit test?
Unless you, @Yuki992411, are up for the task to add a test?

**Comment by Yuki992411** [comment]
Thank you for your PR approval.
This is my first time to write test code.
However, it is a very good opportunity, so I will create a unit test myself and add it to issue #4569.
If you find any mistakes, please give me feedback.

File: test/ml/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+set(DEFAULT ON)
+set(build TRUE)
+PCL_SUBSYS_OPTION(build "${SUBSYS_NAME}" "${SUBSYS_DESC}" ${DEFAULT} "${REASON}")
+PCL_SUBSYS_DEPEND(build "${SUBSYS_NAME}" DEPS ${SUBSYS_DEPS} OPT_DEPS ${OPT_DEPS})
+
+if(NOT build)
+  return()
+endif()
+
+PCL_ADD_TEST(ml_kmeans test_ml_kmeans
```
Comment (larshg) [inline test/ml/CMakeLists.txt:14]: Please make it a single line and add an additional empty line to the of file (required in multiple files).

**Comment by larshg** [comment]
You are missing:

```
/* ---[ */
int
main (int argc, char** argv)
{
  testing::InitGoogleTest (&argc, argv);
  return (RUN_ALL_TESTS ());
}

```

In the end of the test file.

**Comment by larshg** [comment]
And you can uses the short spdx version of the copyright:
```
/*
 * SPDX-License-Identifier: BSD-3-Clause
 *
 *  Point Cloud Library (PCL) - www.pointclouds.org
 *  Copyright (c) 2020-, Open Perception
 *
 *  All rights reserved
 */
```

beside the small changes, its looking good!

File: test/ml/test_kmeans.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+#include <pcl/test/gtest.h>
+#include <pcl/ml/kmeans.h>
+
+using namespace pcl;
+
+using Point = std::vector<float>;
+
+// Use mersenne twister for generating random value
+// std::random_device seed_gen;
+std::mt19937 engine ( 2021 );
```
Comment (larshg) [inline test/ml/test_kmeans.cpp:50]: Please make use of this random:
https://github.com/PointCloudLibrary/pcl/blob/master/common/include/pcl/common/random.h

File: test/ml/CMakeLists.txt (hunk 1 above)
Comment (Yuki992411) [reply test/ml/CMakeLists.txt:14]: Which one should I rewrite? 
```
PCL_ADD_TEST(ml_kmeans test_ml_kmeans FILES test_kmeans.cpp LINK_WITH pcl_gtest pcl_ml)
```
or
```
PCL_ADD_TEST(ml_kmeans test_ml_kmeans
            FILES test_kmeans.cpp LINK_WITH pcl_gtest pcl_ml)
```

Comment (larshg) [reply test/ml/CMakeLists.txt:14]: To the first - a single line.
Thats how they are in common at least.

File: test/ml/test_kmeans.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+/* ---[ */
+int
+main (int argc, char** argv)
+{
+  testing::InitGoogleTest (&argc, argv);
+  return (RUN_ALL_TESTS ());
+}
+/* ]--- */
```
Comment (larshg) [inline test/ml/test_kmeans.cpp:105]: Still missing a empty line here I think.

Comment (Yuki992411) [reply test/ml/test_kmeans.cpp:105]: Oh, I'm sorry.
I missed it. Overwrite the previous commit.

Comment (larshg) [reply test/ml/test_kmeans.cpp]: No reason to be sorry, its just part of the reviewing 😄

**Comment by larshg** [comment]
Looking good. Lets wait for CI's and @mvieth review on the unit test.
If no further changes are required, you can do a interactive rebase and set two of respective commits as fixups.

Then we have one commit with correction and one with the addition of unit test.

**Comment by Yuki992411** [comment]
OK!
Thank you for first review 😁

**Review by mvieth — APPROVED** [review]
Just two minor suggestions:
1. Switch the Case1/2 and Kmeans_computeCentroids. Usually it goes from [general to specififc](https://github.com/google/googletest/blob/master/docs/primer.md#simple-tests)
2. We could also e.g. test if the number of computed centroids is as requested, i.e. 9 and 1.

**Comment by Yuki992411** [comment]
Thank you for your suggestion and review. @larshg @mvieth 
In addition to the above suggestions, I have made some corrections.

* According to [primer.md](https://github.com/google/googletest/blob/master/docs/primer.md#simple-tests) it was written that the ```TEST()``` arguments should not contain any underscores (```_```). So, I changed ```Kmeans_computeCentroids``` to ```ComputeCentroids```.
* I fixed public indent at line 23.
* I summarized commits only in fix code and unit tests.

Please check my commit.

**Review by mvieth — APPROVED** [review]
Very nice, thanks again.
@larshg do you want to have a last look at the changes?

**Comment by larshg** [comment]
Looks ready 🚀

**Pull request merged**